Major refactoring of the Maven dependency/dependents resolution

proksch opened the pull request:
The old implementation of the Maven dependency/dependents resolution had some major shortcomings and did not support many Maven features that affect dependency resolution. This rework contains several tasks:

- I have improved the interface of the resolution engine to make it easier to integrate
- I have replaced the existing logic for Maven dependency/dependents resolution with a version that supports the most important* resolution features of Maven.
- I have combined all Maven related data structures in this repository (some were part of the `data-processing` repo before)
- I have overhauled the data serialization that allows to store an intermediate version of the resolver locally to skip the necessary DB queries on re-instantiation
- I made the resolver engine incremental, to make it possible to feed it through Kafka messages without the need of frequent restarts
- In addition to the DB-based resolver implementation, I have provided a REST-based one that we can use as soon as the latest changes are being merged in the data-processing repository and the micro-service is up and running in the environment

(*) Currently, a sat-based solution for version constraints is missing, as is the handling of the "import scope" (when dependencyManagement sections are imported through dependencies to "poms" -- normal dependency management sections are fully supported)

ashkboos commented on core/src/test/java/eu/fasten/core/maven/data/DependencyTest.java:
```diff
@@ -15,25 +15,285 @@
+    }
+
+    @Test
+    public void equalitDiffVersionsSet() {
```
Maybe a typo in function names from here onwards, equalit -> equality?

proksch replied:
yes, fixed.

ashkboos commented on pom.xml:
```diff
@@ -180,6 +180,8 @@
                 <version>${maven.compiler.plugin.version}</version>
                 <configuration>
                     <release>11</release>
+                    <compilerArgument>-g:vars</compilerArgument>
```
What do you use these arguments for?

proksch replied:
The idea is to preserve variable names in method signatures so they can be used by tooling. I came across this in the context of Jackson modules. I am not 100% yet whether this is working or how to test it. Also does not hurt though. ;)

proksch replied:
Admittedly, with what I have learned about plugin management recently, I might be that we need to move this into the `pluginManagement` section.

ashkboos reviewed (commented):
Amir and I have some minor comments that you could clarify. Overall looks good to us. We appreciate it if you can merge the PR.

proksch commented on core/src/main/java/eu/fasten/core/maven/data/PackageVersionNotFoundException.java:
```diff
@@ -1,24 +1,24 @@
 
 public class PackageVersionNotFoundException extends RuntimeException {
+
+    private static final long serialVersionUID = 4776812947859392189L;
```
This is just the canonical way to get around a static analysis warning... `RuntimeException` implements `Serializable`, so it should declare that field.

proksch commented on core/src/main/java/eu/fasten/core/maven/data/VersionConstraint.java:
```diff
@@ -15,103 +15,154 @@
+        if (isRange()) {
+            String asd = spec.substring(1, spec.length() - 1);
+            var parts = asd.split(",", -1);
+            return parts[1].isEmpty() ? "999" : parts[1];
```
1. Oh, I think the `asd` was just left-over when I wanted to debug that part... I'll inline that. xD
2. "999" Just means "very high"

proksch commented on core/src/main/java/eu/fasten/core/maven/data/VersionConstraint.java:
```diff
@@ -15,103 +15,154 @@
+    }
+
+    private static int find(String hay, int idx, char... needles) {
+        var found = false;
```
Nice catch. Fixed.

proksch commented on core/src/main/java/eu/fasten/core/maven/resolution/MavenDependencyResolver.java:
```diff
@@ -0,0 +1,267 @@
+                throw new MavenResolutionException(
+                        String.format("Cannot find coordinate %s:%s:%s", parts[0], parts[1], parts[2]));
+            } else {
+                return resolve(config, new HashSet<>(), QueueData.startFrom(pom));
```
The variable is just used for preventing an endless loop. I do not think that an assignment would make it easier to read.

ashkboos replied:
Yeah, that's true, it may not make it easier to read. But I usually try to lower the number of arguments whenever possible as a good practice, I think it is more because of maintainability and usability of the method rather than readability.

proksch replied:
No, it is not possible to leave it out, as it is handed down in the recursive calls.

proksch commented on core/src/main/java/eu/fasten/core/maven/resolution/MavenDependencyResolver.java:
```diff
@@ -0,0 +1,267 @@
+        return depSet;
+    }
+
+    private static ResolvedRevision toRR(QueueData data) {
```
done

proksch commented on core/src/main/java/eu/fasten/core/maven/runners/MyRunner.java:
```diff
@@ -0,0 +1,63 @@
+import eu.fasten.core.maven.resolution.ResolverConfig;
+
+public class MyRunner {
+
```
I would like to migrate all classes in that package to the data-processing repo, which is why I kept it for now... It just felt out of scope for now, I would like to postpone that to the next PR.

proksch commented on core/src/test/java/eu/f4sten/test/HttpTestServer.java:
```diff
@@ -0,0 +1,208 @@
+import org.apache.http.client.utils.URIBuilder;
+
+// DO NOT EDIT THIS FILE, THIS IS A CLONE FROM `data-processing/test-utils`
+public class HttpTestServer {
```
I am not sure yet how to handle this best.

proksch commented on core/src/test/java/eu/fasten/core/utils/AssertsTest.java:
```diff
@@ -15,57 +15,84 @@
-		});
-	}
+    @Test
+    public void assertContains_fail() {
```
I am only using it in tests from time to time as some kind of a qualifier to describe what the test is about.

proksch commented on core/src/main/java/eu/fasten/core/maven/resolution/RestMavenResolver.java:
```diff
@@ -0,0 +1,88 @@
+import jakarta.ws.rs.core.GenericType;
+import jakarta.ws.rs.ext.ContextResolver;
+
+public class RestMavenResolver implements IMavenResolver {
```
That's the class that can be used to query the micro service.

proksch commented on core/src/main/java/eu/fasten/core/maven/resolution/MavenDependentsResolver.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    private static boolean matchesScope(Scope dep, Scope request, boolean alwaysIncludeProvided,
+            boolean isTransitiveDep) {
```
Good catch. I removed it.

proksch commented:
I have addressed all review comments and implemented the ideas regarding immutability that we have discussed in our pair programming session.

I saw that you have already approved the PR... before we merge it though, I would like to do one final check on Monster to see whether the implementation does scale to the current size of our DB.

ashkboos commented on core/src/main/java/eu/fasten/core/maven/data/Dependency.java:
```diff
@@ -15,235 +15,329 @@
+        return ToStringBuilder.reflectionToString(this, ToStringStyle.MULTI_LINE_STYLE);
+    }
+
+    // TODO delete everything below
```
Do you still want to do this TODO?

proksch replied:
This is just a TODO for the future. At some point, the remaining code in that class should be removed by refactoring the JSON handling. This is out-of-scope for this PR though.

ashkboos commented on core/src/main/java/eu/fasten/core/maven/data/Exclusion.java:
```diff
@@ -17,47 +17,73 @@
+        return String.format("%s:%s", groupId, artifactId);
+    }
+
+    // TODO remove everything below
```
What about this one?

proksch replied:
This is just a TODO for the future. At some point, the remaining code in that class should be removed by refactoring the JSON handling. This is out-of-scope for this PR though.

ashkboos commented on core/src/main/java/eu/fasten/core/maven/data/Pom.java:
```diff
@@ -0,0 +1,266 @@
+        if (dependencies == null || dependencies.isEmpty()) {
+            this.dependencies = NO_DEPS;
+        } else {
+            // TODO check for (and prevent) "double wrapping"
```
Could you please clarify this TODO?

proksch replied:
This is just a TODO for the future. I have a potential optimization in mind, but this is out-of-scope for this PR.

proksch commented:
After extensive testing on `monster`, it seems that the resolution is ready to go live. The `data-processing` part that includes the micro-service deployment will follow in the next days.
